crates/ruff/src/rules/ruff/rules/unreachable.rs

```diff
@@ -0,0 +1,964 @@
+            {
+                return None;
+            }
+
+            let is_set = (self.bits[index] & (1 << shift)) != 0;
+            shift += 1;
+            if !is_set {
+                return Some((index * usize::BITS as usize) + shift - 1);
+            }
+        })
+    }
+}
+
+/// Set bits in `reached_map` for all blocks that are reached in `blocks`
+/// starting with block at index `idx`.
+fn mark_reached(reached_map: &mut Bitmap, blocks: &[BasicBlock<'_>], mut idx: BlockIndex) {
+    loop {
+        let block = &blocks[idx];
+        if reached_map.set(idx) {
+            return; // Block already visited, no needed to do it again.
+        }
+
+        match &block.next {
+            NextBlock::Always(next) => idx = *next,
+            NextBlock::If {
+                condition,
+                next,
+                orelse,
+            } => {
+                match taken(condition) {
+                    Some(true) => idx = *next,    // Always taken.
+                    Some(false) => idx = *orelse, // Never taken.
+                    None => {
+                        // Don't know, both branches might be taken.
+                        idx = *next;
+                        mark_reached(reached_map, blocks, *orelse);
+                    }
+                }
+            }
+            NextBlock::Terminate => return,
+        }
+    }
+}
+
+/// Determines if `condition` is taken.
+/// Returns `Some(true)` if the condition is always true, e.g. `if True`, same
+/// with `Some(false)` if it's never taken. If it can't be determined it returns
+/// `None`, e.g. `If i == 100`.
+fn taken(condition: &Condition) -> Option<bool> {
+    // TODO: add more cases to this where we can determine a condition
+    // statically. For now we only consider constant booleans.
+    match condition {
+        Condition::Test(expr) => match expr {
+            Expr::Constant(constant) => constant.value.as_bool().copied(),
+            _ => None,
+        },
+        Condition::Iterator(_) => None,
+        Condition::Match { .. } => None,
+    }
+}
+
+/// Collection of basic block.
+#[derive(Debug, PartialEq)]
+struct BasicBlocks<'stmt> {
+    /// # Notes
+    ///
+    /// The order of these block is unspecified. However it's guaranteed that
+    /// the last block is the statement in the function and the first block is
+    /// the last statement. The block are more or less in reverse order, but it
+    /// gets fussy around control flow statements (e.g. `if` statements).
+    ///
+    /// For loop blocks, and similar recurring control flows, the end of the
+    /// body will point to the loop block again (to create the loop). However an
+    /// oddity here is that this block might contain statements before the loop
+    /// itself which, of course, won't be executed again.
+    ///
+    /// For example:
+    /// ```python
+    /// i = 0          # block 0
+    /// while True:    #
```

Comment:
What's the reason for creating a connection to `block 0` vs creating four blocks: 

1. Everything before the while `i = 0`
2. While header (`while True:`)
3. While body
4. After the body

And connecting

* 1 and 2 with an unconditional jump
* 2 and 3 with a conditional jump
* 2 and 4 with a conditional jump
* 3 and 2 with an unconditional jump

Reply:
For the purposes of detecting unreachable code, your block 1 doesn't contain any control flow, notwithstanding function calls that always raise an exception, thus block 2 is always reached when block 1 is reached. Creating one block instead of two is then simply less work for us later on. Rustc does the same thing, see https://rustc-dev-guide.rust-lang.org/appendix/background.html#what-is-a-control-flow-graph.

Other then the addition block, it's what we roughly create at the moment, see https://github.com/astral-sh/ruff/pull/5384/files#diff-a98f67bee97e7c459fad838467a76c7cbcd7c66beca1bae4826d1fa4054c4e5d (crates/ruff/src/rules/ruff/rules/snapshots/ruff__rules__ruff__rules__unreachable__tests__while.py_12.snap) for a graph of code that is quite similar.